### ISSUE TITLE
Incorrect path matching for WebSocket using PathMappings

### ISSUE THREAD

**Issue by lachlan-roberts** [eclipse/jetty.project#4776]
**Description**
Currently we use `PathMappings` to do URI mappings to websocket server endpoints. However this does not use the process described in JSR356 and it seems it gives incorrect matches for some cases.

Example:
```java
PathMappings<String> mapping = new PathMappings<>();
mapping.put(new UriTemplatePathSpec("/{a}/b"), "endpointA");
mapping.put(new UriTemplatePathSpec("/{a}/{b}"), "endpointB");
assertThat(mapping.getMatch("/a/b").getResource(), is("endpointA"));
```

As described in the JSR356 we should traverse the segments left to right, and prefer exact matches to those that have variable segments. Because of this we should match `"endpointA"` in this example as its last segment is an exact match, however `PathMappings` currently matches `"endpointB"`.

Paragraph from JSR356 describing the matching process:
```
Since the endpoint paths are either relative URIs or URI templates level 1, the paths do not match if they do
not have the same number of segments, using ’/’ as the separator. So, the container will traverse the segments
of the endpoint paths with the same number of segments as the incoming URI from left to right, comparing
each segment with the corresponding segment of the incoming URI. At each segment, the implementation
will retain those endpoint paths that match exactly, or if there are none, those that are a variable segment,
before moving to check the next segment. If there is an endpoint path at the end of this process there is a
match.
```

**Comment by gregw** [comment]
@lachlan-roberts let's chat about possible solutions to this before you commit too much code

**Comment by joakime** [comment]
Is this something that's tested on the websocket TCK?

**Comment by joakime** [comment]
> @lachlan-roberts let's chat about possible solutions to this before you commit too much code

I'd like to be involved in this discussion.

**Comment by gregw** [comment]
just in case I'm not about when you start discussing this.....  my thoughts are as follows:
 + I believe this is a TCK test
 + there are a bunch of examples in the spec that we should implement in a unit test first as proof we are not following the spec... plus a copy of the TCK test (which I think is the one in the description).
 + The broad groups of prefix glob, middle glob and suffix glob have the relative priorities from the servlet path spec, which is not the same as the websocket one.    So something fundamental probably needs to change.
 + I like having a single path matching class that can match any type of pattern... even if we have few cases of mixed matching.  Thus I really don't want to split out the matching.
 + If we do have to split out the matching, we should at least keep a common core matcher class and have pluggable strategies that support webapp, websocket and/or regex matching.

**Comment by joakime** [comment]
See also last discussion about path matching at websocket-api -> https://github.com/eclipse-ee4j/websocket-api/issues/170

**Comment by lachlan-roberts** [comment]
fixed with PR #4884
